File: impl/ocaml/nix/emit.ml
(** Print an AST as valid Nix SOURCE.

    Not to be confused with {!Printer}, which reproduces
    [nix-instantiate --parse]'s debug form for differential testing. This one
    emits `.nix` a human can read and Nix can evaluate, and it is the
    transpiler: the arrow [EXPR -> .nix] in [docs/architecture.md].

    {1 Fully parenthesised, on purpose}

    Every compound expression is wrapped. That is uglier than necessary and it
    is the right first move: extra parentheses cannot change the parse, so the
    emitted file is correct by construction rather than correct if the
    precedence table was transcribed properly. `docs/nix-internals.md` records
    two precedence levels that surprise everyone ([!] binds looser than [+],
    [//] tighter than the comparisons); a minimal-parenthesis printer has to
    get both right, and this one cannot get them wrong.

    Making the output pretty is a separate, later job with its own test: emit,
    re-parse, and compare the ASTs. *)

open Ast

let buf_add = Buffer.add_string

let escape s =
  let b = Buffer.create (String.length s + 2) in
  String.iter
    (fun c ->
      match c with
      | '"' -> Buffer.add_string b "\\\""
      | '\\' -> Buffer.add_string b "\\\\"
      | '\n' -> Buffer.add_string b "\\n"
      | '\r' -> Buffer.add_string b "\\r"
      | '\t' -> Buffer.add_string b "\\t"
      (* `${` in a literal must be escaped or Nix reads an interpolation. *)
      | '$' -> Buffer.add_string b "\\$"
      | c -> Buffer.add_char b c)
    s ;
  Buffer.contents b

let op_text = function
  | Add -> "+"
  | Sub -> "-"
  | Mul -> "*"
  | Div -> "/"
  | Update -> "//"
  | Concat -> "++"
  | Eq -> "=="
  | Neq -> "!="
  | Lt -> "<"
  | Gt -> ">"
  | Le -> "<="
  | Ge -> ">="
  | And -> "&&"
  | Or -> "||"
  | Impl -> "->"

let rec pp b (e : t) =
  match e with
  | Int n -> buf_add b (string_of_int n)
  | Float f -> buf_add b (Printf.sprintf "%g" f)
  | Var x -> buf_add b x
  | Path p -> buf_add b p
  | Search_path p -> buf_add b ("<" ^ p ^ ">")
  | Uri u -> buf_add b ("\"" ^ escape u ^ "\"")
  | Path_interp parts ->
      (* The leading path is absolute after parsing, so this emits an absolute
         interpolated path, which is valid Nix and denotes the same file. *)
      List.iter
        (function
          | Lit s -> buf_add b s
          | Anti (Path p) -> buf_add b p
          | Anti e ->
              buf_add b "${" ;
              pp b e ;
              buf_add b "}")
        parts
  | Str parts -> pp_string b parts
  | Ind_str parts -> pp_string b parts
  | Not e ->
      buf_add b "(!" ;
      pp b e ;
      buf_add b ")"
  | Neg e ->
      buf_add b "(-" ;
      pp b e ;
      buf_add b ")"
  | Op (o, x, y) ->
      buf_add b "(" ;
      pp b x ;
      buf_add b (" " ^ op_text o ^ " ") ;
      pp b y ;
      buf_add b ")"
  | Has_attr (e, path) ->
      buf_add b "(" ;
      pp b e ;
      buf_add b " ? " ;
      pp_attrpath b path ;
      buf_add b ")"
  | Apply (f, a) ->
      buf_add b "(" ;
      pp b f ;
      buf_add b " " ;
      pp b a ;
      buf_add b ")"
  | Select (e, path, dflt) ->
      buf_add b "(" ;
      pp b e ;
      buf_add b "." ;
      pp_attrpath b path ;
      Option.iter
        (fun d ->
          buf_add b " or " ;
          pp b d)
        dflt ;
      buf_add b ")"
  | Lambda (p, body) ->
      buf_add b "(" ;
      pp_pattern b p ;
      buf_add b ": " ;
      pp b body ;
      buf_add b ")"
  | List items ->
      buf_add b "[ " ;
      List.iter
        (fun i ->
          pp b i ;
          buf_add b " ")
        items ;
      buf_add b "]"
  | Attr_set {recursive; binds} ->
      if recursive then buf_add b "rec " ;
      buf_add b "{ " ;
      List.iter (pp_bind b) binds ;
      buf_add b "}"
  | Let (binds, body) ->
      buf_add b "(let " ;
      List.iter (pp_bind b) binds ;
      buf_add b "in " ;
      pp b body ;
      buf_add b ")"
  | With (e, body) ->
      buf_add b "(with " ;
      pp b e ;
      buf_add b "; " ;
      pp b body ;
      buf_add b ")"
  | Assert (c, body) ->
      buf_add b "(assert " ;
      pp b c ;
      buf_add b "; " ;
      pp b body ;
      buf_add b ")"
  | If (c, t, f) ->
      buf_add b "(if " ;
      pp b c ;
      buf_add b " then " ;
      pp b t ;
      buf_add b " else " ;
      pp b f ;
      buf_add b ")"

and pp_string b parts =
  buf_add b "\"" ;
  List.iter
    (fun p ->
      match p with
      | Lit s -> buf_add b (escape s)
      | Anti e ->
          buf_add b "${" ;
          pp b e ;
          buf_add b "}")
    parts ;
  buf_add b "\""

and pp_attr b = function
  | Aid x -> buf_add b x
  | Adyn e ->
      buf_add b "${" ;
      pp b e ;
      buf_add b "}"
  | Astr [Lit s] -> buf_add b ("\"" ^ escape s ^ "\"")
  | Astr parts ->
      buf_add b "${" ;
      pp_string b parts ;
      buf_add b "}"

and pp_attrpath b path =
  List.iteri
    (fun i a ->
      if i > 0 then buf_add b "." ;
      pp_attr b a)
    path

and pp_bind b = function
  | Bind (path, e) ->
      pp_attrpath b path ;
      buf_add b " = " ;
      pp b e ;
      buf_add b "; "
  | Inherit (from, attrs) ->
      buf_add b "inherit" ;
      Option.iter
        (fun e ->
          buf_add b " (" ;
          pp b e ;
          buf_add b ")")
        from ;
      List.iter
        (fun a ->
          buf_add b " " ;
          pp_attr b a)
        attrs ;
      buf_add b "; "

and pp_pattern b = function
  | Pvar x -> buf_add b x
  | Pset {formals; ellipsis; alias} ->
      buf_add b "{ " ;
      List.iteri
        (fun i (name, dflt) ->
          if i > 0 then buf_add b ", " ;
          buf_add b name ;
          Option.iter
            (fun d ->
              buf_add b " ? " ;
              pp b d)
            dflt)
        formals ;
      if ellipsis then buf_add b (if formals = [] then "..." else ", ...") ;
      buf_add b " }" ;
      Option.iter (fun a -> buf_add b (" @ " ^ a)) alias

(** The AST as Nix source. *)
let to_string (e : t) : string =
  let b = Buffer.create 512 in
  pp b e ;
  Buffer.contents b

File: impl/rust/src/nix/emit.rs
//! Print an AST as valid Nix SOURCE. This is the transpiler.
//!
//! The arrow `EXPR -> .nix` of `docs/architecture.md`.
//!
//! Fully parenthesised, on purpose. Extra parentheses cannot change a parse, so
//! the emitted file is correct BY CONSTRUCTION rather than correct if the
//! precedence table was transcribed properly. `docs/nix-internals.md` records
//! two levels that surprise everyone (`!` binds looser than `+`, `//` tighter
//! than the comparisons); a minimal-parenthesis printer has to get both right
//! and this one cannot get them wrong.
//!
//! Making the output pretty is a separate, later job with its own test: emit,
//! re-parse, compare the ASTs.

use std::fmt::Write as _;

use super::ast::{Attr, AttrPath, Binding, Expr, Part, Pattern};

/// Write a path so it lexes back as a PATH.
///
/// The ROOT path cannot be written as a bare separator: that lexes as the
/// DIVISION OPERATOR, so the emitted source does not parse at all. Nix source
/// spells it `/.`, which is what nixpkgs itself writes.
///
/// The differential printer prints it as `/`, because Nix does. The two
/// printers disagree on purpose, and only the round-trip law could see it.
fn path_literal(p: &str) -> &str {
    if p == "/" { "/." } else { p }
}

/// Write a float so it lexes back as a FLOAT.
///
/// The default formatting prints `1.0` as `1`, which re-parses as an INTEGER,
/// and Nix distinguishes the two: `1 / 2` is 0 for integers and 0.5 for floats.
/// So the transpiler was silently able to change arithmetic. Found by the
/// round-trip law, not by the eleven conformance intents, none of which
/// contains a float.
///
/// The differential printer must NOT do this either: `nix-instantiate --parse`
/// really does print `1.0` as `1`, and that is pinned by a vector.
fn float_literal(f: f64) -> String {
    let text = format!("{f}");
    if text.contains(['.', 'e', 'E', 'n']) {
        text
    } else {
        format!("{text}.0")
    }
}

/// Escape a string's contents for a double-quoted Nix literal.
fn escape(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    for c in text.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            // `${` in a literal must be escaped or Nix reads an interpolation.
            '$' => out.push_str("\\$"),
            c => out.push(c),
        }
    }
    out
}

fn parts(out: &mut String, parts: &[Part]) {
    out.push('"');
    for p in parts {
        match p {
            Part::Lit(s) => out.push_str(&escape(s)),
            Part::Anti(e) => {
                out.push_str("${");
                expr(out, e);
                out.push('}');
            }
        }
    }
    out.push('"');
}

fn attr(out: &mut String, a: &Attr) {
    match a {
        Attr::Id(name) => out.push_str(name),
        Attr::Dyn(e) => {
            out.push_str("${");
            expr(out, e);
            out.push('}');
        }
        // A string-named attribute is emitted as a STRING, not wrapped in an
        // interpolation. Writing `${"a${x}b"}` is valid Nix denoting the same
        // attribute, but it re-parses as a DYNAMIC name rather than a string
        // one, and Nix keeps those apart. The differential printer collapses
        // them, so nothing but the round-trip law could see it.
        Attr::Str(ps) => match ps.as_slice() {
            [Part::Lit(s)] => {
                let _ = write!(out, "\"{}\"", escape(s));
            }
            _ => parts(out, ps),
        },
    }
}

fn attrpath(out: &mut String, path: &AttrPath) {
    for (i, a) in path.iter().enumerate() {
        if i > 0 {
            out.push('.');
        }
        attr(out, a);
    }
}

fn binding(out: &mut String, b: &Binding) {
    match b {
        Binding::Bind(path, value) => {
            attrpath(out, path);
            out.push_str(" = ");
            expr(out, value);
            out.push_str("; ");
        }
        Binding::Inherit(source, attrs) => {
            out.push_str("inherit");
            if let Some(e) = source {
                out.push_str(" (");
                expr(out, e);
                out.push(')');
            }
            for a in attrs {
                out.push(' ');
                attr(out, a);
            }
            out.push_str("; ");
        }
    }
}

fn pattern(out: &mut String, p: &Pattern) {
    match p {
        Pattern::Var(name) => out.push_str(name),
        Pattern::Set {
            formals,
            ellipsis,
            alias,
        } => {
            out.push_str("{ ");
            for (i, (name, default)) in formals.iter().enumerate() {
                if i > 0 {
                    out.push_str(", ");
                }
                out.push_str(name);
                if let Some(d) = default {
                    out.push_str(" ? ");
                    expr(out, d);
                }
            }
            if *ellipsis {
                out.push_str(if formals.is_empty() { "..." } else { ", ..." });
            }
            out.push_str(" }");
            if let Some(a) = alias {
                let _ = write!(out, " @ {a}");
            }
        }
    }
}

fn expr(out: &mut String, e: &Expr) {
    match e {
        Expr::Int(n) => {
            let _ = write!(out, "{n}");
        }
        Expr::Float(f) => out.push_str(&float_literal(*f)),
        Expr::Var(x) => out.push_str(x),
        Expr::Path(p) => out.push_str(path_literal(p)),
        Expr::SearchPath(p) => {
            let _ = write!(out, "<{p}>");
        }
        Expr::Uri(u) => {
            let _ = write!(out, "\"{}\"", escape(u));
        }
        Expr::Str(ps) | Expr::IndStr(ps) => parts(out, ps),
        Expr::PathInterp(ps) => {
            // The leading path is absolute after parsing, so this emits an
            // absolute interpolated path: valid Nix, and the same file.
            for p in ps {
                match p {
                    Part::Lit(s) => out.push_str(s),
                    Part::Anti(Expr::Path(path)) => out.push_str(path),
                    Part::Anti(e) => {
                        out.push_str("${");
                        expr(out, e);
                        out.push('}');
                    }
                }
            }
        }
        Expr::Not(e) => {
            out.push_str("(!");
            expr(out, e);
            out.push(')');
        }
        Expr::Neg(e) => {
            out.push_str("(-");
            expr(out, e);
            out.push(')');
        }
        Expr::Op(op, a, b) => {
            out.push('(');
            expr(out, a);
            let _ = write!(out, " {} ", op.text());
            expr(out, b);
            out.push(')');
        }
        Expr::HasAttr(e, path) => {
            out.push('(');
            expr(out, e);
            out.push_str(" ? ");
            attrpath(out, path);
            out.push(')');
        }
        Expr::Apply(f, a) => {
            out.push('(');
            expr(out, f);
            out.push(' ');
            expr(out, a);
            out.push(')');
        }
        Expr::Select(e, path, default) => {
            out.push('(');
            expr(out, e);
            out.push('.');
            attrpath(out, path);
            if let Some(d) = default {
                out.push_str(" or ");
                expr(out, d);
            }
            out.push(')');
        }
        Expr::Lambda(p, body) => {
            out.push('(');
            pattern(out, p);
            out.push_str(": ");
            expr(out, body);
            out.push(')');
        }
        Expr::List(items) => {
            out.push_str("[ ");
            for i in items {
                expr(out, i);
                out.push(' ');
            }
            out.push(']');
        }
        Expr::AttrSet { recursive, binds } => {
            if *recursive {
                out.push_str("rec ");
            }
            out.push_str("{ ");
            for b in binds {
                binding(out, b);
            }
            out.push('}');
        }
        Expr::Let(binds, body) => {
            out.push_str("(let ");
            for b in binds {
                binding(out, b);
            }
            out.push_str("in ");
            expr(out, body);
            out.push(')');
        }
        Expr::With(scope, body) => {
            out.push_str("(with ");
            expr(out, scope);
            out.push_str("; ");
            expr(out, body);
            out.push(')');
        }
        Expr::Assert(c, body) => {
            out.push_str("(assert ");
            expr(out, c);
            out.push_str("; ");
            expr(out, body);
            out.push(')');
        }
        Expr::If(c, t, f) => {
            out.push_str("(if ");
            expr(out, c);
            out.push_str(" then ");
            expr(out, t);
            out.push_str(" else ");
            expr(out, f);
            out.push(')');
        }
    }
}

/// The AST as Nix source.
pub fn to_nix(e: &Expr) -> String {
    let mut out = String::with_capacity(512);
    expr(&mut out, e);
    out
}
